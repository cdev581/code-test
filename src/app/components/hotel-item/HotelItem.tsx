"use client";

import Image from "next/image";
import { useTranslations } from "next-intl";
import Rating from "@/app/components/rating/Rating";
import { Hotel } from "@/types/Hotels.types";
import styles from "./HotelItem.module.scss";

export default function HotelItem({ hotel }: { hotel: Hotel }) {
  const t = useTranslations("hotels");

  const handleItemClick = (event: React.MouseEvent, href: string) => {
    if (!(event.target instanceof HTMLAnchorElement)) {
      window.open(href, "_blank", "noopener,noreferrer");
    }
  };

  return (
    <div className={styles.item} onClick={e => handleItemClick(e, hotel.href)}>
      <div className={styles.depiction}>
        <Image
          src={hotel.image.src}
          alt={hotel.image.alt}
          width={150}
          height={125}
          className={styles.image}
          unoptimized
        />
        <div className={styles.promotion}>{hotel.promotionTitle}</div>
      </div>
      <div className={styles.content}>
        <div className={styles.details}>
          <h3 className={styles.title}>
            <a href={hotel.href} target="_blank" rel="noopener noreferrer">
              {hotel.title}
            </a>
            {hotel.rating && (
              <Rating amount={hotel?.rating?.ratingValue} type={hotel?.rating?.ratingType}>
              </Rating>
            )}
          </h3>
          <div className={styles.address}>{hotel.address}</div>
          <div className={styles.offer}>{hotel.offerName}</div>
          <div className={styles.cancellation}>{t(`cancellation.${hotel.cancellationType}`)}</div>
        </div>
        <div className={styles.pricing}>
          <div className={styles.duration}>
            {t("total")} ({hotel?.price?.currency ?? ""})
          </div>
          <div className={styles.price}><sup>{t("sign")}</sup>{hotel?.price?.amount ?? ""}</div>
          <div className={styles.saving}>
            {hotel.savings && <> {t("save")} {t("sign")}{hotel?.savings?.amount ?? ""}~</>}
          </div>
        </div>
      </div>
    </div>
  );
}