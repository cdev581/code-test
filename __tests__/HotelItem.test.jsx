import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import { NextIntlClientProvider } from "next-intl";
import HotelItem from "../src/app/components/hotel-item/HotelItem";
import hotelsData from "../src/data/hotels.json";
import formatHotelsData from "../src/lib/format-hotels-data";

jest.mock("../src/app/components/loader/PlaneLoader.js", () => jest.fn().mockImplementation(() => {}));

beforeEach(() => {
  global.structuredClone = jest.fn((obj) => JSON.parse(JSON.stringify(obj)));
  global.hotels = global.structuredClone(formatHotelsData(hotelsData));
});

afterEach(() => {
  jest.restoreAllMocks();
});

describe("HotelItem", () => {
  const messages = require(`../messages/en.json`);

  it("renders a hotel item with savings unchanged", async () => {
    const hotel = global.hotels[0];
    hotel.rating = null;

    const { container } = render(
      <NextIntlClientProvider messages={messages} locale="en">
        <HotelItem hotel={hotel} />
      </NextIntlClientProvider>
    );
  
    expect(container).toMatchSnapshot();
  });

  it("renders a hotel item without savings unchanged", async () => {
    const hotel = global.hotels[2];
    hotel.rating = null;

    const { container } = render(
      <NextIntlClientProvider messages={messages} locale="en">
        <HotelItem hotel={hotel} />
      </NextIntlClientProvider>
    );
  
    expect(container).toMatchSnapshot();
  });
});