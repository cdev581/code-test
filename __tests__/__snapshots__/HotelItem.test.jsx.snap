// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HotelItem renders a hotel item with savings unchanged 1`] = `
<div>
  <div
    class="item"
  >
    <div
      class="depiction"
    >
      <img
        alt="Image of Courtyard by Marriott Sydney-North Ryde"
        class="image"
        data-nimg="1"
        decoding="async"
        height="125"
        loading="lazy"
        src="https://unsplash.it/145/125/?random=0"
        style="color: transparent;"
        width="150"
      />
      <div
        class="promotion"
      >
        Exclusive Deal
      </div>
    </div>
    <div
      class="content"
    >
      <div
        class="details"
      >
        <h3
          class="title"
        >
          <a
            href="https://qantas.com.au"
            rel="noopener noreferrer"
            target="_blank"
          >
            Courtyard by Marriott Sydney-North Ryde
          </a>
        </h3>
        <div
          class="address"
        >
          7-11 Talavera Rd, North Ryde
        </div>
        <div
          class="offer"
        >
          Deluxe Balcony Room
        </div>
        <div
          class="cancellation"
        />
      </div>
      <div
        class="pricing"
      >
        <div
          class="duration"
        >
          1 night total
           (
          AUD
          )
        </div>
        <div
          class="price"
        >
          <sup>
            $
          </sup>
          329
        </div>
        <div
          class="saving"
        >
           
          Save
           
          $
          30
          ~
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`HotelItem renders a hotel item without savings unchanged 1`] = `
<div>
  <div
    class="item"
  >
    <div
      class="depiction"
    >
      <img
        alt="Image of property"
        class="image"
        data-nimg="1"
        decoding="async"
        height="125"
        loading="lazy"
        src="https://unsplash.it/145/125/?random=2"
        style="color: transparent;"
        width="150"
      />
      <div
        class="promotion"
      >
        Red Hot
      </div>
    </div>
    <div
      class="content"
    >
      <div
        class="details"
      >
        <h3
          class="title"
        >
          <a
            href="https://qantas.com.au"
            rel="noopener noreferrer"
            target="_blank"
          >
            Rydges World Square Sydney
          </a>
        </h3>
        <div
          class="address"
        >
          389 Pitt St, Sydney
        </div>
        <div
          class="offer"
        >
          Deluxe King Room
        </div>
        <div
          class="cancellation"
        />
      </div>
      <div
        class="pricing"
      >
        <div
          class="duration"
        >
          1 night total
           (
          AUD
          )
        </div>
        <div
          class="price"
        >
          <sup>
            $
          </sup>
          227
        </div>
        <div
          class="saving"
        />
      </div>
    </div>
  </div>
</div>
`;
